File: packages/js-engine/tests/src/context_tests.cpp
#include <fmt/core.h>

#include <catch2/catch_test_macros.hpp>
#include <chrono>
#include <future>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

#include "native_engine/engine.hpp"

// TEST_CASE("test null eval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   auto value = engine->execute(context_name, "undefined");
//   REQUIRE(value->is_null_or_undefined());

//   value = engine->execute(context_name, "const test = null; test;");
//   REQUIRE(value->is_null_or_undefined());

//   delete engine;
// }

// TEST_CASE("test bool eval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   auto value = engine->execute(context_name, "let test = (1 == 1); test;");
//   REQUIRE(value->get_bool_value());

//   value = engine->execute(context_name, "test = (100 == 200); test;");
//   REQUIRE(!value->get_bool_value());

//   delete engine;
// }

// TEST_CASE("test integer eval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   auto value = engine->execute(context_name, "1 + 2;");
//   REQUIRE(value->get_int_value() == 3);

//   delete engine;
// }

// TEST_CASE("test float eval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   auto value = engine->execute(context_name, "10.8 + 2.77;");
//   REQUIRE(value->get_float_value() == 13.57f);

//   delete engine;
// }

// TEST_CASE("test string eval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   auto value = engine->execute(context_name, "'hello' + ' ' + 'world';");
//   REQUIRE(value->get_string_value() == "hello world");

//   delete engine;
// }

// TEST_CASE("test console api", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   engine->execute(context_name, "console.log('hello world');");
//   engine->execute(context_name, "console.info('this message is for informational purposes');");
//   engine->execute(context_name, "console.warn('this is a warning message');");
//   engine->execute(context_name, "console.error('a problem has occurred');");
//   engine->execute(context_name, "console.debug('this is a debugging statement');");

//   delete engine;
// }

// TEST_CASE("test event listeners", "[context]") {
//   std::promise<int> future1;
//   std::promise<int> future2;
//   std::promise<json> future3;
//   std::promise<int> future4;

//   std::function<json(json)> func_1 = [&future1](json args) {
//     future1.set_value(1);
//     return nullptr;
//   };

//   std::function<json(json)> func_2 = [&future2](json args) {
//     future2.set_value(1);
//     return nullptr;
//   };

//   std::function<json(json)> func_3 = [&future3](json args) {
//     future3.set_value(args);
//     return nullptr;
//   };

//   std::function<json(json)> func_4 = [&future4](json args) {
//     future4.set_value(1);
//     return nullptr;
//   };

//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);
//   engine->register_function(context_name, "successCallback", func_1);
//   engine->register_function(context_name, "altSuccessCallback", func_2);
//   engine->register_function(context_name, "successCallbackDetails", func_3);
//   engine->register_function(context_name, "successCallbackFunction", func_4);

//   // setting a basic event listener
//   engine->execute(context_name, "addEventListener('myEvent', () => { successCallback(); });");
//   engine->dispatch_event(context_name, "myEvent", nullptr);

//   REQUIRE(future1.get_future().get() == 1);

//   engine->execute(context_name, "addEventListener('myEvent', () => { altSuccessCallback(); });");
//   engine->dispatch_event(context_name, "myEvent", nullptr);

//   REQUIRE(future2.get_future().get() == 1);

//   // basic event listener with details
//   engine->execute(context_name, "addEventListener('myEventDetails', (details) => { successCallbackDetails(details); });");

//   json details;
//   details["name"] = "John Doe";

//   engine->dispatch_event(context_name, "myEventDetails", details);

//   auto value = future3.get_future().get();

//   REQUIRE(value["name"] == "John Doe");

//   // basic event listener with registered global function as callback arg
//   engine->execute(context_name, "addEventListener('myEventCallback', (resolve) => { resolve() });");

//   json args;
//   json callbacks;

//   callbacks["resolve"] = "__cbr::successCallbackFunction";
//   args["callbacks"] = callbacks;

//   engine->dispatch_event(context_name, "myEventCallback", args);

//   REQUIRE(future4.get_future().get() == 1);

//   delete engine;
// }

// TEST_CASE("test error handling", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->create_context(context_name);

//   try {
//     engine->execute(context_name, "() => { throw new Error('this method has an error'); }();");
//     REQUIRE(false);
//   } catch (JavaScriptException& ex) {
//     std::string error_message = ex.what();
//     REQUIRE(error_message.find("JS Exception") != std::string::npos);
//   }

//   try {
//     engine->execute(context_name, "addEventListener(");
//     REQUIRE(false);
//   } catch (JavaScriptException& ex) {
//     std::string error_message = ex.what();
//     REQUIRE(error_message.find("JS Exception") != std::string::npos);
//   }

//   try {
//     engine->execute(context_name, "addEventListener('myThrowingEvent', () => { throw new Error('this event throws an error') })");
//     engine->dispatch_event(context_name, "myThrowingEvent", nullptr);
//   } catch (JavaScriptException& ex) {
//     std::string error_message = ex.what();
//     REQUIRE(error_message.find("JS Exception") != std::string::npos);
//   }

//   // test handling C++ exceptions
//   std::promise<void> future;

//   std::function<json(json)> func = [&future](json args) {
//     throw std::invalid_argument("a problem occurred in C++");
//     return nullptr;
//   };

//   try {
//     engine->register_function(context_name, "exceptionCallback", func);
//     engine->execute(context_name, "addEventListener('myThrowingJVMEvent', () => { exceptionCallback(); })");
//     engine->dispatch_event(context_name, "myThrowingJVMEvent", nullptr);
//   } catch (JavaScriptException& ex) {
//     std::string error_message = ex.what();
//     REQUIRE(error_message.find("JS Exception") != std::string::npos);
//   }
// }

// TEST_CASE("test set timeout", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->start();

//   engine->create_context(context_name);

//   std::promise<int> promise1;
//   std::promise<int> promise2;

//   std::function<json(json)> func_1 = [&promise1](json args) {
//     promise1.set_value(1);
//     return nullptr;
//   };

//   std::function<json(json)> func_2 = [&promise2](json args) {
//     promise2.set_value(1);
//     return nullptr;
//   };

//   engine->register_function(context_name, "timeoutCallback", func_1);
//   engine->register_function(context_name, "cancelTimeoutCallback", func_2);

//   auto ret = engine->execute(context_name, "setTimeout(() => { timeoutCallback(); }, 2000)");
//   auto timer_id = ret->get_int_value();

//   REQUIRE(timer_id > 0);

//   auto future1 = promise1.get_future();

//   auto status = future1.wait_for(std::chrono::seconds(3));
//   if (status == std::future_status::timeout) {
//     FAIL("timer did not fire");
//   }

//   REQUIRE(future1.get() == 1);

//   ret = engine->execute(context_name, "setTimeout(() => { cancelTimeoutCallback() }, 4000)");
//   timer_id = ret->get_int_value();

//   REQUIRE(timer_id > 0);

//   engine->execute(context_name, fmt::format("clearTimeout({});", timer_id));

//   auto future2 = promise2.get_future();

//   status = future2.wait_for(std::chrono::seconds(5));
//   if (status == std::future_status::timeout) {
//     SUCCEED("timer did not fire after canceled");
//   } else {
//     FAIL("timer was not canceled");
//   }

//   engine->stop();
//   delete engine;
// }

// TEST_CASE("test set interval", "[context]") {
//   std::string context_name = "io.ionic.android_js_engine";
//   auto engine = new Engine();
//   engine->start();

//   engine->create_context(context_name);

//   std::promise<int> promise1;
//   std::promise<int> promise2;

//   int calls = 0;

//   std::function<json(json)> func_1 = [&calls](json args) {
//     calls++;
//     return nullptr;
//   };

//   engine->register_function(context_name, "intervalCallback", func_1);
//   auto ret = engine->execute(context_name, "setInterval(() => { intervalCallback() }, 2000)");
//   auto timer_id = ret->get_int_value();

//   REQUIRE(timer_id > 0);

//   auto future1 = promise1.get_future();
//   auto status = future1.wait_for(std::chrono::seconds(8));
//   if (status == std::future_status::timeout) {
//     REQUIRE(calls == 4);
//   } else {
//     FAIL("???");
//   }

//   engine->execute(context_name, fmt::format("clearInterval({});", timer_id));

//   auto future2 = promise2.get_future();
//   status = future2.wait_for(std::chrono::seconds(3));
//   if (status == std::future_status::timeout) {
//     REQUIRE(calls == 4);
//   } else {
//     FAIL("???");
//   }
//   engine->stop();
//   delete engine;
// }

TEST_CASE("test fetch", "[context]") {
  auto engine = new Engine();

  auto test_thread = std::thread([engine] {
    std::string context_name = "io.ionic.android_js_engine";
    engine->create_context(context_name);

    std::promise<json> success_promise;
    std::promise<json> options_success_promise;

    std::function<json(json)> callback_1 = [&success_promise](json args) {
      success_promise.set_value(args);
      return nullptr;
    };

    std::function<json(json)> callback_2 = [&options_success_promise](json args) {
      options_success_promise.set_value(args);
      return nullptr;
    };

    engine->register_function(context_name, "successCallback", callback_1);
    engine->register_function(context_name, "successCallback2", callback_2);

    std::string basic_fetch_example =
        "fetch('https://jsonplaceholder.typicode.com/todos/1')"
        ".then(response => response.json())"
        ".then(json => { successCallback(json); })"
        ".catch(err => { console.error(err); });";

    engine->execute(context_name, basic_fetch_example);

    auto future1 = success_promise.get_future();
    auto status = future1.wait_for(std::chrono::seconds(5));
    if (status == std::future_status::timeout) {
      FAIL("could not get request");
    }

    auto response_data = future1.get();
    auto title = response_data["title"].template get<std::string>();
    REQUIRE(title == "delectus aut autem");

    std::string basic_fetch_options_example =
        "fetch('https://jsonplaceholder.typicode.com/posts', {"
        "method: 'POST',"
        "body: JSON.stringify({"
        "title: 'foo',"
        "body: 'bar',"
        "userId: 1,"
        "}),"
        "headers: {"
        "'Content-type': 'application/json; charset=UTF-8',"
        "}"
        "})"
        ".catch(err => { console.error(err); })"
        ".then(response => response.json())"
        ".then(json => { successCallback2(json); })";

    engine->execute(context_name, basic_fetch_options_example);

    auto future2 = options_success_promise.get_future();
    status = future2.wait_for(std::chrono::seconds(5));
    if (status == std::future_status::timeout) {
      FAIL("could not get request");
    }

    response_data = future2.get();
    auto body = response_data["body"].template get<std::string>();
    REQUIRE(body == "bar");

    engine->stop();
  });

  engine->start();
  test_thread.join();

  delete engine;
}

File: packages/capacitor-plugin/ios/Tests/RunnerEngineTests/ContextTests.swift
// swiftlint:disable all

import XCTest
import JavaScriptCore
@testable import RunnerEngine

final class ContextTests: XCTestCase {
    func testNullEval() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.nulleval")
        guard let value = try context.execute(code: "undefined") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isUndefined)

        guard let value = try context.execute(code: "const test = null; test;") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isNull)
    }

    func testBoolEval() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.booleval")

        guard let value = try context.execute(code: "let test = (1 == 1); test;") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isBoolean)
        XCTAssertTrue(value.toBool())

        guard let value = try context.execute(code: "test = (100 == 200); test;") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isBoolean)
        XCTAssertFalse(value.toBool())
    }

    func testIntegerEval() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.inteval")

        guard let value = try context.execute(code: "1 +2;") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isNumber)
        XCTAssertEqual(3, value.toInt32())
    }

    func testFloatEval() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.dbleval")

        guard let value = try context.execute(code: "10.8 + 2.77;") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isNumber)
        XCTAssertEqual(13.57, value.toDouble())
    }

    func testStringEval() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.stringeval")

        guard let value = try context.execute(code: "'hello' + ' ' + 'world';") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isString)
        XCTAssertEqual("hello world", value.toString())
    }

    func testConsoleAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.consoletests")

        _ = try context.execute(code: "console.log('hello world');")
        _ = try context.execute(code: "console.info('this message is for informational purposes');")
        _ = try context.execute(code: "console.warn('this is a warning message');")
        _ = try context.execute(code: "console.error('a problem has occurred');")
        _ = try context.execute(code: "console.debug('this is a debugging statement');")
    }

    func testEventListeners() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.eventlistenertests")

        let expectation = XCTestExpectation(description: "Run callback on event")
        let expectation2 = XCTestExpectation(description: "Run another callback on event")
        let expectation3 = XCTestExpectation(description: "Run another callback on event with details")

        var successCallbackCalled = 0
        var altSuccessCallbackCalled = 0

        let successCallback: @convention(block) () -> Void = {
            successCallbackCalled += 1
            expectation.fulfill()
        }
        let successCallback2: @convention(block) () -> Void = {
            altSuccessCallbackCalled += 1
            expectation2.fulfill()
        }

        let successCallbackDetails: @convention(block) (JSValue) -> Void = {details in
            XCTAssertTrue(details.isObject)

            let detailsObj = details.toDictionary() as? [String: String]
            XCTAssertEqual(detailsObj?["name"], "John Doe")

            expectation3.fulfill()
        }

        context.jsContext.setObject(successCallback, forKeyedSubscript: "successCallback" as NSString)
        context.jsContext.setObject(successCallback2, forKeyedSubscript: "altSuccessCallback" as NSString)
        context.jsContext.setObject(successCallbackDetails, forKeyedSubscript: "successCallbackDetails" as NSString)

        _ = try context.execute(code: "addEventListener('myEvent', () => { successCallback(); });")
        try  context.dispatchEvent(event: "myEvent")

        wait(for: [expectation], timeout: 1)

        _ = try context.execute(code: "addEventListener('myEvent', () => { altSuccessCallback(); });")
        try context.dispatchEvent(event: "myEvent")

        wait(for: [expectation2], timeout: 1)

        XCTAssertEqual(successCallbackCalled, 1)
        XCTAssertEqual(altSuccessCallbackCalled, 1)

        var detailsObject = [String: String]()
        detailsObject["name"] = "John Doe"

        _ = try context.execute(code: "addEventListener('myEventDetails', (details) => { successCallbackDetails(details); });")
        try context.dispatchEvent(event: "myEventDetails", details: detailsObject)

        wait(for: [expectation3], timeout: 1)
    }

    func testCryptoAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.cryptotests")

        guard let value = try context.execute(code: "const array = new Uint32Array(10);  crypto.getRandomValues(array); array;") else {
            XCTFail("value is null")
            return
        }
        if value.isNull || value.isUndefined {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isObject)
        XCTAssertEqual(10, value.toDictionary().count)

        guard let value = try context.execute(code: "crypto.randomUUID();") else {
            XCTFail("value is null")
            return
        }
        if value.isNull || value.isUndefined {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isString)
        XCTAssertEqual(36, value.toString().count)

    }

    func testTimeoutAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.timeouttests")

        let expectation = XCTestExpectation(description: "Run callback set on timeout")
        let cancelExpectation = XCTestExpectation(description: "Cancel an already set up timeout")

        let successCallback: @convention(block) () -> Void = {
            expectation.fulfill()
        }

        let failureCallback: @convention(block) () -> Void = {
            XCTFail("callback was called and not canceled")
        }

        context.jsContext.setObject(successCallback, forKeyedSubscript: "successCallback" as NSString)
        context.jsContext.setObject(failureCallback, forKeyedSubscript: "failureCallback" as NSString)

        guard let value = try context.execute(code: "setTimeout(() => { successCallback(); }, 2000)") else {
            XCTFail("value is null")
            return
        }

        XCTAssertGreaterThan(value.toInt32(), 0)

        wait(for: [expectation], timeout: 2.5)

        guard let value = try context.execute(code: "setTimeout(() => { failureCallback(); }, 4000)") else {
            XCTFail("value is null")
            return
        }

        let timerId = value.toInt32()

        _ = Timer.scheduledTimer(withTimeInterval: TimeInterval(4.5), repeats: false) {_ in
            cancelExpectation.fulfill()
        }

        _ = try context.execute(code: "clearTimeout(\(timerId))")

        wait(for: [cancelExpectation], timeout: 5)

    }

    func testIntervalAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.intervaltests")

        var calls = 0

        let expectation = XCTestExpectation(description: "Run callback set on timeout")

        let timerCallback: @convention(block) () -> Void = {
            calls += 1
        }

        context.jsContext.setObject(timerCallback, forKeyedSubscript: "timerCallback" as NSString)

        guard let value = try context.execute(code: "setInterval(() => { timerCallback(); }, 2000)") else {
            XCTFail("value is null")
            return
        }

        let timerId = value.toInt32()

        _ = Timer.scheduledTimer(withTimeInterval: TimeInterval(8), repeats: false) {_ in
            do {
                _ = try context.execute(code: "clearInterval(\(timerId))")
            } catch {
                XCTFail("\(error)")
            }
        }

        _ = Timer.scheduledTimer(withTimeInterval: TimeInterval(11), repeats: false) {_ in
            expectation.fulfill()
        }

        wait(for: [expectation], timeout: 12)

        XCTAssertEqual(calls, 4)

    }

    func testTextEncoderAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.textencodertests")

        guard let value = try context.execute(code: "const encoder = new TextEncoder(); encoder.encode('€');") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isObject)

        guard let arrayObject = value.toDictionary() as? [String: Int] else {
            XCTFail("could not get array object")
            return
        }
        XCTAssertEqual(226, arrayObject["0"])
        XCTAssertEqual(130, arrayObject["1"])
        XCTAssertEqual(172, arrayObject["2"])
    }

    func testTextDecoderAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.textdecodertests")

        guard let value = try context.execute(code: "const win1251decoder = new TextDecoder(\"windows-1251\"); win1251decoder.decode(new Uint8Array([ 207, 240, 232, 226, 229, 242, 44, 32, 236, 232, 240, 33]));") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isString)
        XCTAssertEqual("Привет, мир!", value.toString())

        guard let value = try context.execute(code: "const decoder = new TextDecoder(); decoder.decode(new Uint8Array([240, 160, 174, 183]));") else {
            XCTFail("value is null")
            return
        }

        XCTAssertTrue(value.isString)
        XCTAssertEqual("𠮷", value.toString())
    }

    func testErrorHandling() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.testexceptions")

        XCTAssertThrowsError(try context.execute(code: "() => { throw new Error('this method has an error'); }();")) { error in
            XCTAssertTrue(String(describing: error).contains("SyntaxError: Unexpected token"))
        }

        XCTAssertThrowsError(try context.execute(code: "addEventListener(")) { error in
            XCTAssertTrue(String(describing: error).contains("SyntaxError: Unexpected end of script"))
        }

        _ = try context.execute(code: "addEventListener('myThrowingEvent', () => { throw new Error('this event throws an error') })")

        XCTAssertThrowsError(try context.dispatchEvent(event: "myThrowingEvent")) { error in
            XCTAssertTrue(String(describing: error).contains("this event throws an error"))
        }
    }

    func testFetchAPI() throws {
        let runner = Runner()
        let context = try runner.createContext(name: "io.backgroundrunner.testfetch")

        let expectation = XCTestExpectation(description: "Run callback on completed (or failed) fetch")
        let optionsExpectation = XCTestExpectation(description: "Run callback on completed options fetch")
        let failureExpectation = XCTestExpectation(description: "Run catch callback on failed fetch")

        let successCallback: @convention(block) () -> Void = {
            expectation.fulfill()
        }

        let failureCallback: @convention(block) () -> Void = {
            failureExpectation.fulfill()
        }

        let optionsSuccessCallback: @convention(block) () -> Void = {
            optionsExpectation.fulfill()
        }

        context.jsContext.setObject(successCallback, forKeyedSubscript: "successCallback" as NSString)
        context.jsContext.setObject(optionsSuccessCallback, forKeyedSubscript: "successCallback2" as NSString)
        context.jsContext.setObject(failureCallback, forKeyedSubscript: "failureCallback" as NSString)

        let basicFetchExample = """
            fetch('https://jsonplaceholder.typicode.com/todos/1')
                .then(response => response.json())
                .then(json => { console.log(JSON.stringify(json)); successCallback(); })
                .catch(err => { console.error(err);  successCallback(); });
            """

        _ = try context.execute(code: basicFetchExample)

        wait(for: [expectation], timeout: 3)

        let fetchFailureExample = """
            fetch('https://blablabla.fake/todos/1')
                .catch(err => { console.error(err);  failureCallback(); });
            """

        _ = try context.execute(code: fetchFailureExample)

        wait(for: [failureExpectation], timeout: 3)

        let fetchWithOptionsExample = """
            fetch('https://jsonplaceholder.typicode.com/posts', {
                method: 'POST',
                body: JSON.stringify({
                    title: 'foo',
                    body: 'bar',
                    userId: 1,
                }),
                headers: {
                    'Content-type': 'application/json; charset=UTF-8',
                }
            })
            .catch(err => { console.error(err); })
            .then(response => response.json())
            .then(json => { console.log(JSON.stringify(json)); successCallback2(); })
            """

        _ = try context.execute(code: fetchWithOptionsExample)

        wait(for: [optionsExpectation], timeout: 3)

    }
}

// swiftlint:enable all
